### ISSUE TITLE
Fixed Visual Studio project files, need help with pull request.

### ISSUE THREAD

**Issue by BarabasGitHub** [munificent/wren#270]
There were some files missing from the visual studio project files, so it wouldn't compile properly. (I got linker errors.)

Now I'd like to send a pull request, but I also did something else. (See #269) And I can't manage to make separate pull requests for them. Sorry I'm new here. ^^

**Comment by munificent** [comment]
> Sorry I'm new here. ^^

That's totally fine. Git is pretty far from easy to understand.

I think you want to get your fork set up like this:

1. The master branch is "clean". It tracks the master branch of the main Wren repo and contains no other commits.

2. You have a branch for the VS proj changes.

3. You have a branch for the computed goto changes.

Let's walk through it. I'm assuming your terminal is in the directory for your repo and you're currently on the master branch.

Before we clean up master, we'll make a temporary branch so we don't lose track of the changes you made.

    $ git branch old-master

This makes a new branch named old-master pointing to the same commit that your master branch is pointing to. Now let's get master back to a happy place.

    $ git reset --hard a114f34a2a3532de6149291e994a2d238a58e687

This rolls back your master branch to point to the last commit before any of your changes. Now your master is clean, it's just a bit out of date. Next, we'll pull in the latest changes from my repo. First, run this:

    $ git remote -v

Does it output:

    upstream  https://github.com/munificent/wren (fetch)
    upstream  https://github.com/munificent/wren (push)

If not, do this:

    git remote add upstream https://github.com/munificent/wren

This adds a new remote named "upstream" whose URL is my repo. That lets you pull changes from it. Now we'll do that:

    git fetch upstream

This pulls all of the latest commits from my repo but doesn't make any of your branches point to them. They're just sitting idle in your `.git` directory. Now we'll update your local master with those commits:

    git merge upstream/master

If everything goes right, the output should say "Fast forward". That's good. It means your master is clean and git can update it by just pointing it to the latest commit it pulled from my repo.

Part 1 is done! Huzzah! Now lets set up feature branches for your two changes.

    git branch vs-proj
    git checkout vs-proj

This creates a new branch "vs-proj" and switches to it. Right now, it's identical to master. Now let's bring that VS project commit into it:

    git cherry-pick 1fd7f3f2eec8141dfadc991aece9a04cedcb3f09

This takes the commit you made earlier with the project changes and applies it to this branch. Now we'll do the other.

    git checkout master
    git branch goto
    git checkout goto

We make another feature branch. We switch back to master first so that the branch is based on master and not the vs-proj branch. If we didn't do that, this branch would end up with *both* the VS project stuff and the computed goto stuff.

    git cherry-pick 08c4c920361ae81c2a997ae49c721c95c4768c78

We apply the computed goto commit onto the new goto feature branch. Now that the feature branches are good, we can remove our temporary branch:

    git branch -D old-master

Swell! Your local repo is all squared away now. Now let's get this all up to GitHub. That's just:

    git push origin master
    git push origin vs-proj
    git push origin goto

I *think* I have this right, but I can't test all of it locally since I don't own your fork. Try that out and see how it goes. If everything goes well, you can then see your two feature branches on GitHub. From there, you can click the little button to create a pull request for each one.

**Comment by BarabasGitHub** [comment]
WOW, that really doesn't sound easy. Thanks a lot for the walkthrough! It almost worked. The only thing it complains about is when I try to push the master branch: git push origin master
It said it couldn't push it because it was behind it's remote counterpart. I used --force-with-lease and now it looks like it did the job.

**Comment by bjorn** [comment]
I'll just try to make life a bit easier by mentioning some shortcuts. Maybe it's easier without them, you decide! :-)

Here we add an `upstream` remote and then fetch from it: 
```
git remote add upstream https://github.com/munificent/wren
git fetch upstream
```
Since fetching from a remote after adding it is so common, there's a shortcut for it:
```
git remote add -f upstream https://github.com/munificent/wren
```

The following makes a branch called 'goto' that starts out from 'master':
```
git checkout master
git branch goto
git checkout goto
```
but it is equivalent to:
```
git checkout -b goto master
```

Here we're pushing multiple branches to origin:
```
git push origin master
git push origin vs-proj
git push origin goto
```
but we can push them all at once like:
```
git push origin master vs-proj goto
```

Initially I think git is hard to understand, but after a few months you'll no longer want to use anything else. :-)

**Comment by BarabasGitHub** [comment]
Thanks for the tips =)
I'm used to hg, and then actually only via the TortoiseHg GUI. I use svn at work, but there I don't do much more than update and commit. :-P

**Comment by MarcoLizza** [comment]
If you are used to UI-based source control tools you can also try and use Atlassian's SourceTree. It's quite easy to use with GitHub, too...

**Comment by Rohansi** [comment]
I fixed this in #312.
